toplevel/topdirs.ml

```diff
@@ -225,39 +225,42 @@ let printer_type ppf typename =
   printer_type
 
 let match_simple_printer_type desc printer_type =
-  Ctype.begin_def();
-  let ty_arg = Ctype.newvar() in
-  begin try
-    Ctype.unify !toplevel_env
-      (Ctype.newconstr printer_type [ty_arg])
-      (Ctype.instance desc.val_type);
-  with Ctype.Unify _ ->
-    raise Bad_printing_function
-  end;
-  Ctype.end_def();
-  Ctype.generalize ty_arg;
+  let ty_arg =
+    Ctype.wrap_def ~post:Ctype.generalize begin fun () ->
```

Comment:
Note to self: this would read nicer with `let@ () = Ctype.wrap_def ~post:Ctype.generalize in ...` -- see #9887.